crux/src/Crux/Goal.hs

```diff
@@ -389,48 +392,81 @@ proveGoalsOnline bak opts _ctxt explainFailure (Just gs0) =
 
   hasUnsatCores = unsatCores opts && not (yicesMCSat opts)
 
+  howManyAbducts = case (getNAbducts opts) of Just n  -> n
+                                              Nothing -> 0
+
   failfast = proofGoalsFailFast opts
 
   go (start,end) sp assumptionsInScope gn gs nameMap = do
+    -- traverse goal tree
     case gs of
-
+      -- case: assumption in context for all the contained goals
       Assuming asms gs1 ->
         do ps <- flattenAssumptions sym asms
            forM_ ps $ \asm ->
              unless (trivialAssumption asm) $
+               -- extract predicate from assumption
                do let p = assumptionPred asm
+                  -- create formula, assert to SMT solver, create new name and add to nameMap
                   nm <- doAssume =<< mkFormula conn p
                   bindName nm (Left asm) nameMap
+           -- recursive call
            res <- go (start,end) sp (assumptionsInScope <> asms) gn gs1 nameMap
            return (Assuming (mconcat (map singleAssumption ps)) res)
-
+      -- case: proof obligation in the context of all previously-made assumptions
       Prove p ->
+        -- number of processed goals gives goal number to prove
         do goalNumber <- totalProcessedGoals <$> readIORef gn
            start goalNumber
-           nm <- doAssume =<< mkFormula conn =<< notPred sym (p ^. labeledPred)
+           -- negate goal, create formula
+           t <- mkFormula conn =<< notPred sym (p ^. labeledPred)
+           -- assert formula to SMT solver, create new name and add to nameMap
+           -- This is done in a new assertion frame if abduction is turned on
+           if (howManyAbducts /= 0) then
+             inNewFrame2Open sp
```

Comment:
I'm a bit alarmed by the fact that we open and close the `2` frame in separate steps, as I worry that it will be easy to forget to close the frame if we refactor the code further. Indeed, the logic for closing the frame is already somewhat complex, as we have to remember to close the frame _only_ if abduction is enabled.

I would prefer to first define a combinator that runs an action within a newly pushed frame 2, popping it after the action completes. Something like `what4`'s [`inNewFrame`](https://github.com/GaloisInc/what4/blob/6f5e0fe9bef58234603ccf6914c32ea1ba2f9766/what4/src/What4/Protocol/Online.hs#L362-L364), but perhaps not as complicated. I'm thinking:

```hs
inNewFrame2 :: (MonadIO m, MonadMask m, SMTReadWriter solver) => SolverProcess scope solver -> m a -> m a
inNewFrame2 sp action = do
  inNewFrame2Open sp
  val <- action
  inNewFrame2Close sp
  return val
```

Then, the code here might look something like this:

```hs
let withFrame2 sp = if howManyAbducts /= 0 then inNewFrame2 sp else id
withFrame2 sp $ do
  nm <- doAssume t
  ...
```

Note that `what4`'s `inNewFrame` function is a bit more complicated than the implementation of `inNewFrame2` I propose above, since it also makes an effort to be exception-safe. For now, however, I think it suffices to define a simplistic `inNewFrame2` function locally here. We can open an issue to upstream this code into `what4` (and making it exception-safe) later.

Reply:
This was done in f2b4b8351325b422fc611e2b479bef886c77fbb0. I've opened GaloisInc/what4#218 as a reminder to consider upstreaming `inNewFrame2` to `what4`, so it would be worth citing that issue in within the `TODO` that I added in `inNewFrame2`'s comments.